### PR TITLE
[TimeSeries] Implement Notebook comparing metrics with Pandas

### PR DESCRIPTION
**Pull request by dsdinter**

### CODE REVIEW

File: timeseries-streaming/timeseries-java-applications/Examples/build.gradle [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+// Enable code formatting
+spotless {
+    java {
+        licenseHeaderFile '../Apache_Licence_For_Spotless'
+        googleJavaFormat('1.7')
+    }
+}
+
+task forex_example(type: JavaExec) {
```
Comment (rezarokni) [inline timeseries-streaming/timeseries-java-applications/Examples/build.gradle:58]: I dont think we have done this yet, but maybe we should look at adding notes on any tasks that we add. What they do etc...

File: timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexBatchPipeline.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+     */
+    ExampleForexPipelineOptions options =
+            PipelineOptionsFactory.fromArgs(args).as(ExampleForexPipelineOptions.class);
+    options.setAppName("ForexExample");
+
+
+    LocalDateTime ldt = LocalDateTime.parse("2020-05-12T00:00:00");
+    String timezone = options.getTimezone();
+    if (timezone == null) {
+      throw new IllegalArgumentException("Please specify timezone parameter");
```
Comment (rezarokni) [inline timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexBatchPipeline.java:78]: Notes, on why they need timezone and which timezone, the viewer the data etc?

File: timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexBatchPipeline.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+    long millis = zdt.toInstant().toEpochMilli();
+
+    options.setTypeOneComputationsLengthInSecs(options.getResampleSec());
+    options.setTypeTwoComputationsLengthInSecs(options.getWindowSec());
+    options.setSequenceLengthInSeconds(options.getWindowSec());
+
+    options.setAbsoluteStopTimeMSTimestamp(millis);
+    options.setEnableHoldAndPropogate(TRUE);
+    // We fill gaps for 10 min during quiet periods of forex markets
```
Comment (rezarokni) [inline timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexBatchPipeline.java:90]: I think 10 mins maybe too small. Maybe something more in the day range?

File: timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexBatchPipeline.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    }
+    ZonedDateTime zdt = ldt.atZone(ZoneId.of(timezone));
+
+    long millis = zdt.toInstant().toEpochMilli();
+
+    options.setTypeOneComputationsLengthInSecs(options.getResampleSec());
+    options.setTypeTwoComputationsLengthInSecs(options.getWindowSec());
+    options.setSequenceLengthInSeconds(options.getWindowSec());
+
+    options.setAbsoluteStopTimeMSTimestamp(millis);
```
Comment (rezarokni) [inline timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexBatchPipeline.java:88]: Notes on why we have both absolute time and ttl here would be great.

File: timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexBatchPipeline.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    options.setEnableHoldAndPropogate(TRUE);
+    // We fill gaps for 10 min during quiet periods of forex markets
+    options.setTTLDurationSecs(600);
+
+    Pipeline p = Pipeline.create(options);
+
+//    String resourceName = "EURUSD-2020-05-11_2020-05-11.csv";
```
Comment (rezarokni) [inline timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexBatchPipeline.java:95]: Remove?

File: timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexBatchPipeline.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+//      e.printStackTrace();
+//    }
+//    String absolutePath = null;
+//    try {
+//      absolutePath = file.getAbsolutePath();
+//    } catch (Exception e) {
+//      e.printStackTrace();
+//    }
+
+    String absolutePath = "gs://4f93039d-7a71-48e7-800b-59f642741ebd/MarketData/FX/EURUSD/tick/EURUSD-2020-05-11_2020-05-11.csv";
```
Comment (rezarokni) [inline timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexBatchPipeline.java:111]: Is this a public folder?

File: timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexBatchPipeline.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+    /**
+     * ***********************************************************************************************************
+     * The data has only one key, to allow the type 1 computations to be done in parallel we set the
+     * {@link GenerateComputations#hotKeyFanOut()}
+     * ***********************************************************************************************************
+     */
+    GenerateComputations.Builder generateComputations = null;
+    if (options.getMetrics() == null) {
```
Comment (rezarokni) [inline timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexBatchPipeline.java:120]: What happens if getMetrics is not null? Would be good to comment.

File: timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ExampleForexPipelineOptions.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+ * limitations under the License.
+ */
+package com.google.dataflow.sample.timeseriesflow.examples.fsi.forex;
+
+import com.google.dataflow.sample.timeseriesflow.TFXOptions;
+import com.google.dataflow.sample.timeseriesflow.TimeseriesStreamingOptions;
+import org.apache.beam.sdk.annotations.Experimental;
+
+@Experimental
+public interface ExampleForexPipelineOptions extends TimeseriesStreamingOptions {
```
Comment (rezarokni) [inline timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ExampleForexPipelineOptions.java:25]: Comments for the options.

File: timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+import org.apache.beam.sdk.schemas.annotations.SchemaFieldName;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollection;
+import org.joda.time.Instant;
+
+import static jdk.nashorn.internal.objects.Global.print;
+
+public class ForexCSVAdaptor {
```
Comment (rezarokni) [inline timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java:48]: Would be good to comment about this is to be used for back testing or bootstrap, normally we expect to be in Stream mode.

File: timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+
+      ConvertCSVForex convertCSVForex;
+
+      public ParseCSV(ConvertCSVForex convertCSVForex) {
+        this.convertCSVForex = convertCSVForex;
+      }
+
+      ObjectReader objectReader;
+
+      @StartBundle
```
Comment (rezarokni) [inline timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java:81]: Would this also worin @Setup ?

File: timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+                // OUTPUT Bid volume
+                outputReceiver.outputWithTimestamp(
+                    TSDataPoint.newBuilder()
+                        .setKey(key.toBuilder().setMinorKeyString("BID_VOLUME"))
+                        .setTimestamp(Timestamps.fromMillis(millis))
+                        .setData(CommonUtils.createNumData(forex.bid_volume))
+                        .build(), Instant.ofEpochMilli(millis));
+              }
+            } catch (Exception ex) {
+              System.out.println(String.format("Unable to parse record: %s", ex));
```
Comment (rezarokni) [inline timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java:142]: Consider adding deadletter support. 
Also use LOGGER for messages when inside the transform.

File: timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+                // Setting UTC using Reykjavik time zone
+                ZonedDateTime zdt = ldt.atZone(ZoneId.of("Atlantic/Reykjavik"));
+
+                long millis = zdt.toInstant().toEpochMilli();
+
+                // OUTPUT Ask Price
+                outputReceiver.outputWithTimestamp(
+                    TSDataPoint.newBuilder()
+                        .setKey(key.toBuilder().setMinorKeyString("ASK"))
+                        .setTimestamp(Timestamps.fromMillis(millis))
```
Comment (rezarokni) [inline timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java:113]: Might be work creating the timestamps object once and then assign to all these.

File: timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+        }
+      }
+    }
+  }
+
+  /**
+   * ticker|timestamp|ask|bid|ask_volume|bid_volume Intermediary POJO used for conversion of CSV
+   * file.
+   */
+  @DefaultSchema(JavaFieldSchema.class)
```
Comment (rezarokni) [inline timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java:157]: Do we expect all these to always be not null?

File: timeseries-streaming/timeseries-java-applications/TimeSeriesPipeline/src/main/java/com/google/dataflow/sample/timeseriesflow/verifier/TSDataPointVerifier.java [hunk 14]
```diff
@@ -61,6 +63,11 @@ public void process(@Element TSDataPoint input, OutputReceiver<TSDataPoint> o) {
             String.format("TSDataPoint has no timestamp. %s", input.toString()));
       }
 
+      if (!(time.isAfter(GlobalWindow.TIMESTAMP_MIN_VALUE) && time.isBefore(GlobalWindow.TIMESTAMP_MAX_VALUE))) {
```
Comment (rezarokni) [inline timeseries-streaming/timeseries-java-applications/TimeSeriesPipeline/src/main/java/com/google/dataflow/sample/timeseriesflow/verifier/TSDataPointVerifier.java:66]: Maybe better in a different commit?

**Review by rezarokni — CHANGES_REQUESTED** [review]

File: timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java (hunk 9 above)
Comment (dsdinter) [reply timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java:46]: Done

**Review by dsdinter — COMMENTED** [review]
Just reviewed some of the comments and applied the suggestions, will do some refactoring as second round to include the rest of the other suggestions. Thanks for the feedback!

File: timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java [hunk 15]
```diff
@@ -45,6 +45,7 @@
 
 import static jdk.nashorn.internal.objects.Global.print;
 
+// This class is used for backtesting or bootstrap in batch mode, we would normally implement for streaming workloads
```
Comment (rezarokni) [inline timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java]: This class is used for backtesting or bootstrap in batch mode. The primary usage of the timeseries library being for live stream mode processing.

File: timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexBatchPipeline.java [hunk 16]
```diff
@@ -85,44 +85,28 @@ public static void main(String[] args) {
...
     options.setSequenceLengthInSeconds(options.getWindowSec());
 
+    // Absolute time is required for batch jobs to determine when to stop filling gaps in bound datasets
     options.setAbsoluteStopTimeMSTimestamp(millis);
     options.setEnableHoldAndPropogate(TRUE);
-    // We fill gaps for 10 min during quiet periods of forex markets
-    options.setTTLDurationSecs(600);
+
+    // We fill gaps for 24 hours during quiet periods of forex markets, before reaching the absolute stop time
+    options.setTTLDurationSecs(86400);
```
[the remark this answers is not in the document]
Comment (dsdinter) [reply timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexBatchPipeline.java:67]: Good point, should be 49 hours, from Friday night when US closes (9pm UTC), until Sunday night (10pm UTC), when Asian markets open again, I will round up to 50 hours and comment the reason.

File: timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java (hunk 10 above)
Comment (dsdinter) [reply timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java]: Done.

File: timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java (hunk 11 above)
Comment (dsdinter) [reply timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java]: Done, added basic deadletter support.

File: timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java (hunk 13 above)
Comment (dsdinter) [reply timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java:169]: I would say yes, but the logic includes now checks if some of the values failed to be converted, sending these elements to the deadletter queue.

File: timeseries-streaming/timeseries-java-applications/TimeSeriesPipeline/src/main/java/com/google/dataflow/sample/timeseriesflow/verifier/TSDataPointVerifier.java (hunk 14 above)
Comment (dsdinter) [reply timeseries-streaming/timeseries-java-applications/TimeSeriesPipeline/src/main/java/com/google/dataflow/sample/timeseriesflow/verifier/TSDataPointVerifier.java]: Sure, I reverted the commit and will apply in a separate PR.

**Comment by dsdinter** [comment]
@rezarokni change requests applied, let me know if it looks good to merge please.

**Review by dsdinter — COMMENTED** [review]
All reviewed, feel free to review and approve.

**Review by rezarokni — APPROVED** [review]

**Pull request merged**